### ISSUE TITLE
Cannot require gem if BUNDLE_PATH is set

### ISSUE THREAD

**Issue by pocke** [rubygems/rubygems#3298]
I found the problem while developing RuboCop.
https://github.com/rubocop-hq/rubocop/issues/7124

# Problem

Ruby does not find a gem that is installed by bundler with `BUNDLE_PATH` since ruby/ruby@8f37629 . This commit updated bundler code in ruby/ruby repository.

# Reproducing

```dockerfile
FROM ubuntu:18.04

ENV BUNDLER_MERGE_COMMIT 8f37629519ad330032a38ac0e871b2912ed38a1b
ENV WORKING_DIR /work

RUN apt-get update -y
RUN apt-get install -y autoconf bison build-essential libssl-dev libyaml-dev libreadline6-dev zlib1g-dev libncurses5-dev libffi-dev libgdbm5 libgdbm-dev git ruby
RUN git clone --depth=100 https://github.com/ruby/ruby

RUN mkdir $WORKING_DIR && echo "source 'https://rubygems.org'; gem 'pry'" > "$WORKING_DIR/Gemfile"

RUN mkdir /before
RUN mkdir /after

RUN cd ruby/ && git checkout "${BUNDLER_MERGE_COMMIT}~" && autoreconf && ./configure --prefix=/before && make -j && make install
RUN cd ruby/ && git checkout "${BUNDLER_MERGE_COMMIT}"  && autoreconf && ./configure --prefix=/after  && make -j && make install

RUN echo 'Modify me to remove cache'

ENV GEM_HOME /usr/local/bundle
ENV BUNDLE_PATH="$GEM_HOME"

RUN echo "After merge bundler"
RUN cd $WORKING_DIR && /after/bin/bundle install && /after/bin/ruby -v -r pry -e '' || true

RUN echo "Before merge bundler"
RUN cd $WORKING_DIR && /before/bin/bundle install && /before/bin/ruby -v -r pry -e '' || true
```

I use a Dockerfile to reproduce.
It dispplays `cannot load such file -- pry` .

```console
$ docker build .
(snip)

Step 15/18 : RUN echo "After merge bundler"
 ---> Running in 2cd4e88549b1
After merge bundler
Removing intermediate container 2cd4e88549b1
 ---> 5903c6252630
Step 16/18 : RUN cd $WORKING_DIR && /after/bin/bundle install && /after/bin/ruby -v -r pry -e '' || true
 ---> Running in fbfa538b00ea
Don't run Bundler as root. Bundler can ask for sudo if it is needed, and
installing your bundle as root will break this application for all non-root
users on this machine.
Fetching gem metadata from https://rubygems.org/.........
Resolving dependencies...
Using bundler 2.1.0.pre.1
Fetching coderay 1.1.2
Installing coderay 1.1.2
Fetching method_source 0.9.2
Installing method_source 0.9.2
Fetching pry 0.12.2
Installing pry 0.12.2
Bundle complete! 1 Gemfile dependency, 4 gems now installed.
Bundled gems are installed into `/usr/local/bundle`
ruby 2.7.0dev (2019-06-09T12:44:10+09:00 trunk 8f37629519) [x86_64-linux]
/after/lib/ruby/2.7.0/rubygems/core_ext/kernel_require.rb:54:in `require': cannot load such file -- pry (LoadError)
        from /after/lib/ruby/2.7.0/rubygems/core_ext/kernel_require.rb:54:in `require'
Removing intermediate container fbfa538b00ea
 ---> 2c0821585bda
Step 17/18 : RUN echo "Before merge bundler"
 ---> Running in 976aee50b0c0
Before merge bundler
Removing intermediate container 976aee50b0c0
 ---> 16582b269efa
Step 18/18 : RUN cd $WORKING_DIR && /before/bin/bundle install && /before/bin/ruby -v -r pry -e '' || true
 ---> Running in f11a252a1f4d
Don't run Bundler as root. Bundler can ask for sudo if it is needed, and
installing your bundle as root will break this application for all non-root
users on this machine.
Fetching gem metadata from https://rubygems.org/.........
Using bundler 2.1.0.pre.1
Fetching coderay 1.1.2
Installing coderay 1.1.2
Fetching method_source 0.9.2
Installing method_source 0.9.2
Fetching pry 0.12.2
Installing pry 0.12.2
Bundle complete! 1 Gemfile dependency, 4 gems now installed.
Bundled gems are installed into `/usr/local/bundle`
ruby 2.7.0dev (2019-06-09T12:31:12+09:00 trunk 6650899248) [x86_64-linux]
Removing intermediate container f11a252a1f4d
 ---> 5257e284fe9e
Successfully built 5257e284fe9e
```

**Comment by philomory** [comment]
Is there a suggestion for how to work around this in the meantime? I'm using the `ruby:2.7-alpine` Docker image as my base image, and after my code runs `Bundler.setup`, I'm completely unable to load any gems; for example, I see this:

```
/opt/app # bundle info aws-sdk-dynamodb
  * aws-sdk-dynamodb (1.48.0)
        Summary: AWS SDK for Ruby - DynamoDB
        Homepage: https://github.com/aws/aws-sdk-ruby
        Path: /usr/local/bundle/gems/aws-sdk-dynamodb-1.48.0
/opt/app # ruby -e 'require "bundler/setup"; p $LOAD_PATH; require "aws-sdk-dynamodb"'
["/usr/local/lib/ruby/gems/2.7.0/gems/bundler-2.1.4/lib", "/usr/local/lib/ruby/site_ruby/2.7.0", "/usr/local/lib/ruby/site_ruby/2.7.0/x86_64-linux-musl", "/usr/local/lib/ruby/site_ruby", "/usr/local/lib/ruby/vendor_ruby/2.7.0", "/usr/local/lib/ruby/vendor_ruby/2.7.0/x86_64-linux-musl", "/usr/local/lib/ruby/vendor_ruby", "/usr/local/lib/ruby/2.7.0", "/usr/local/lib/ruby/2.7.0/x86_64-linux-musl"]
Traceback (most recent call last):
        1: from -e:1:in `<main>'
-e:1:in `require': cannot load such file -- aws-sdk-dynamodb (LoadError)
```

If I add the following to the top of my Dockerfile:
```
ENV BUNDLE_PATH=
ENV BUNDLE_PATH__SYSTEM=true
```
... it changes the installation path to e.g. `/ruby/2.7.0/gems/aws-sdk-dynamodb-1.48.0`, but, the $LOAD_PATH after `Bundler.setup` is still just as broken as before.

**Comment by akostadinov** [comment]
@deivid-rodriguez , when `BUNDLE_PATH` is set in global config or environment variable causes Bundler NOT to append ruby scope, while setting this option on command line or in project config cause it to DO append that. Just in case that matters for your fix.

You can see more about what I found at https://stackoverflow.com/a/64163511/520567

**Comment by deivid-rodriguez** [comment]
Hei, I lost track of this issue, but https://github.com/rubygems/rubygems/pull/3933 removed the special logic you point out in the stackoverflow thread, so it might've fixed this as a side effect. Could you try if this is reproducible on the latest default branch.

**Comment by deivid-rodriguez** [comment]
I believe this particular issue has been fixed, but I opened #4025 for a separate problem discovered while investigating this. I'll close this now.
